homeassistant/components/forked_daapd/config_flow.py

```diff
@@ -0,0 +1,173 @@
+                    vol.Optional(
+                        CONF_PIPE_CONTROL,
+                        default=self.config_entry.options.get(CONF_PIPE_CONTROL),
+                    ): str,  # https://github.com/home-assistant/core/issues/32819
+                    vol.Optional(
+                        CONF_PIPE_CONTROL_PORT,
+                        default=self.config_entry.options.get(
+                            CONF_PIPE_CONTROL_PORT, DEFAULT_PIPE_CONTROL_PORT
+                        ),
+                    ): int,
+                }
+            ),
+        )
+
+
+def fill_in_schema_dict(some_input):
+    """Fill in schema dict defaults from user_input."""
+    schema_dict = {}
+    for field, _type in DATA_SCHEMA_DICT.items():
+        if some_input.get(str(field)):
+            schema_dict[
+                vol.Optional(str(field), default=some_input[str(field)])
+            ] = _type
+        else:
+            schema_dict[field] = _type
+    return schema_dict
+
+
+class ForkedDaapdFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a forked-daapd config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize."""
+        self.discovery_schema = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Return options flow handler."""
+        return ForkedDaapdOptionsFlowHandler(config_entry)
+
+    async def validate_input(self, user_input):
+        """Validate the user input."""
+        websession = async_get_clientsession(self.hass)
+        validate_result = await ForkedDaapdAPI.test_connection(
+            websession=websession,
+            host=user_input[CONF_HOST],
+            port=user_input[CONF_PORT],
+            password=user_input[CONF_PASSWORD],
+        )
+        return TEST_CONNECTION_ERROR_DICT.get(validate_result, "unknown_error")
+
+    async def async_step_user(self, user_input=None):
+        """Handle a forked-daapd config flow start.
+
+        Manage device specific parameters.
+        """
+        if user_input is not None:
+            validate_result = await self.validate_input(user_input)
+            if validate_result == "ok":  # success
+                _LOGGER.debug("Connected successfully. Creating entry")
+                return self.async_create_entry(
+                    title=f"{FD_NAME} server", data=user_input
+                )
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(fill_in_schema_dict(user_input)),
+                errors={"base": validate_result},
+            )
+        if self.discovery_schema:  # stop at form to allow user to set up manually
+            return self.async_show_form(
+                step_id="user", data_schema=self.discovery_schema, errors={}
+            )
+        await self.async_set_unique_id(CONFIG_FLOW_UNIQUE_ID)  # only allow one entry
```

Comment:
@balloob is it ok to set a constant unique id of only one entry is allowed?

Reply:
It is allowed and we do it for flows that need 1 per network (Sonos, Cast etc). But I don't see a reason in this case why it's limited to a single entry? We have a host so we could connect to multiple ?

Reply:
Looking at the screenshot that @escoand posted, it seems like we can get a unique ID from the zeroconf discovery (the E545… part):

![image](https://user-images.githubusercontent.com/1444314/81122047-f017ca00-8ee4-11ea-8243-db06fe05933d.png)

In that case we can use the `zeroconf` step to add unique IDs to existing entries that have been created manually. This is fine since unique ID for config flow is currently only used to be able to ignore discoveries anyway.

See `tradfri` for an example of how this is done:

https://github.com/home-assistant/core/blob/dev/homeassistant/components/tradfri/config_flow.py#L92-L102

Reply:
Thanks for the feedback.
I originally had it as 1 server per network because in my understanding people only use 1 server, similar to Sonos. That said, there's no reason we can't try to support multiple servers using unique_ids.
However, the _dacp._tcp advertisement posted is not the correct service for detecting a forked-daapd server. There are several services advertised by the server, but the _dacp._tcp one does not seem to include enough information to distinguish a forked-daapd server from other different types of iTunes servers.
The _daap._tcp advertisement which was used does have a Machine ID field which is similar but distinct from the DbId (E545..) field in the _dacp._tcp advertisement. However, looking at the forked-daapd source, both the Machine ID and DbId fields seem to be different hashes (djb hash vs murmur hash) of the library name, so they would not presumably not be unique.
Any suggestions on what else to use for a unique_id?

Reply:
Since there do not seem to be any true unique ids candidates, it seems we either:
1) Go ahead with the current way which only allows 1 server and just revisit this if anyone requests support for multiple servers.
2) Use the library name or Machine ID as the unique_id and just understand that this is not truly unique and would be problematic if two servers have the exact same library name or if an existing library name gets changed.
Let me know which one you guys choose and I'll put the necessary changes through.